Auth Role Checks: Part 4 - Implement user-facing flow changes

dvdchr opened the pull request:
Resolves #3963 
Depends on #4492 and #4665

## Summary

This is the **final** part out of 4 parts that make up the changes related to role authentication (for more details, refer to #3963). This PR introduces **user-facing changes** as it modifies the application flow. Here's a summary of this change: 

### 1. Check role eligibility after selecting a store

With this change, a blocking role eligibility check is performed after user taps "Continue" in Store Picker. If the role eligibility check fails, the user will be presented with a role error page. To see how the error page looks, refer to #4478.

At this moment, there are 3 ways to present a Store Picker: from the Authentication flow, from app launch, or from the Settings page. Here's how the role error page behaves on each of the behaviors:
- **From the authentication flow** 🔒 : The role error page will be directly pushed onto the navigation stack. This is possible because the authentication flow directly pushes Store Picker onto the navigation controller.
- **From app launch** 🚀 : The role error page will be directly pushed onto the _top-most_ navigation controller. In this flow, the Store Picker is wrapped in a navigation controller and then modally presented from a navigation controller. This is why we need to find the top-most navigation controller, and present it from there.
- **From the Settings page** ⚙️ : Same as previous, the role error page will be directly pushed onto the top-most navigation controller. The Store Picker is also presented modally, but the difference is that the Settings page presents it using default presentation style (i.e. it can be dismissed).

In all of the cases, the role eligibility check is performed _before_ the selected `storeID` is updated, therefore user can freely go back from the error page to the Store Picker page to (for example) try accessing a different store. 

Additionally, in the case where the user selected a store and were presented with the role error page, if they restart the app, they would enter the Store Picker again (through the app launch 🚀  flow) – giving them a chance to select a different store.

### 2. Check role eligibility on every app launch

This change also introduces a non-blocking role eligibility check during app launch. The result of the check is stored in `GeneralAppSettings`, and will be consumed by `AppCoordinator` on the next launch, should the result differ from the current state.

The role check is only performed for users that have **authenticated AND selected a store**. The purpose of doing this on every app launch is to detect when the user's roles are updated _after_ they gained access to the store (in the WooCommerce app).

Different from the Store Picker behavior, the error page presented from the app launch flow ⚠️ **may block the user** ⚠️. This can only happen under these conditions:
- User is authenticated,
- Has selected a store,
- User was previously eligible, but the latest role eligibility check failed.

In cases where the API request fails (i.e. we don't know if user is eligible or not), the implementation similarly follows the [implementation on Android](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2002#issuecomment-843736770):

- **If user is previously NOT eligible** and the API call fails, the app will assume that the user is still not eligible. User can retry the eligibility check from the error screen, and if it succeeded, they will be directed to the store.
- **If the user is previously eligible** and the API call fails, the app will assume that the user is still eligible. Although it's not ideal, we don't know if the user is eligible or not because the API call failed. If they turned out to be NOT eligible, they'll get failing API requests on the store due to insufficient permissions.

## Related PRs
- #4473 
- #4478 
- #4619
- #4620 (integration branch)
- #4492
- #4665 

## To Test

Since this involves role modifications, my recommendation is to add a user that you can easily modify in your store. Having a WP-Admin dashboard opened and ready is also handy to quickly apply role changes.

### Testing the flow

### Store Picker: Authentication Flow 🔒 
Make sure the app is in logged out state.

- Login, and enter the Store Picker.
- Select the store where your role is ineligible, and tap "Continue". 
- Verify that the role error page is shown.
- Tap on the "Retry" button. Verify that a notice is shown.
- Tap back. Verify that you are now back at the Store Picker page.
- Enter the Role error page again, and tap Help. Verify that the help form is shown.
- Tap Login with Another Account. Verify that the authentication flow is restarted.
- Go through the authentication process again until you've entered the role error page.
- From the WP Admin, update your role to be eligible.
- Tapping Retry should bring you into the store.

### Store Picker: App Launch Flow 🚀 
Make sure the app is in authenticated state, but has no default store yet.
- Select the store where your role is ineligible, and tap "Continue". 
- Verify that the role error page is shown.
- Tap on the "Retry" button. Verify that a notice is shown.
- Tap back. Verify that you are now back at the Store Picker page.
- Enter the Role error page again, and tap Help. Verify that the help form is shown.
- From the WP Admin, update your role to be eligible.
- Tapping Retry should bring you into the store.

### Store Picker: Settings Flow ⚙️ 
Make sure the app is in authenticated state, and has a default store selected. Also make sure that the user has access to >1 stores.
- Go to Settings and select Switch Store.
- Select the store where the user has ineligible roles, and verify that the role error page is shown.
- Tap on Login With Another Account, verify that the authentication flow is restarted.
- Go to the Store Picker again from Settings, and enter the role error page.
- From the WP Admin, update your role to be eligible.
- Tapping Retry should bring you into the store.

### App Launch Flow

#### Scenario 1: Not Eligible -> Eligible state
Make sure the app is in authenticated state, has a default store selected, and the user is NOT eligible to access.

- Close the app.
- Update the user role to be eligible.
- Open the app.
- Verify that the error page is shown.
- Tap on the Retry button.
- Verify that the main UI is shown.

#### Scenario 2: Eligible -> Not Eligible state
Make sure the app is in authenticated state, has a default store selected, and the user is eligible to access.

- Close the app.
- Update the user role to be NOT eligible.
- Open the app.
- Verify that the error page is shown, not long after the main UI is shown.

#### Scenario 3: Eligible -> Eligible state
Make sure the app is in authenticated state, has a default store selected, and the user is eligible to access.

- Close the app.
- Update the user role to something else, but still an eligible role.
- Open the app.
- Verify that the main UI is shown, and there are no disruptions in the user flow.

#### Scenario 4: NOT Eligible -> NOT Eligible state
Make sure the app is in authenticated state, has a default store selected, and the user is NOT eligible to access.

- Close the app.
- Update the user role to something else, but still an ineligible role.
- Open the app.
- Verify that the role error page is shown.

### Unit tests

- Make sure all the tests are 🟢 .

## Author Checklist

- [x] I have added unit tests where necessary.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dvdchr committed:
Check role eligibility on app launch

dvdchr committed:
Implement role error flow in AppCoordinator

If the user is authenticated and has a default store, AppCoordinator will now check whether there are previous eligibility errors. If so, then the user will be presented with the role error page.

Note that in the case of unexpected errors (i.e. connection loss), the logic will stick with whatever result was stored locally. In the case that the user has been granted the correct roles but got stuck in the error page – there is a retry button to re-check user eligibility.

dvdchr committed:
Add capability to show notice from role error page

dvdchr committed:
Update lastErrorInfo to read directly from user defaults

dvdchr committed:
Adjust RoleErrorViewModel usage in AppCoordinator

dvdchr committed:
Implement role error flow in store picker

dvdchr committed:
Add minor docs, and remove unused code.

dvdchr committed:
Fix unit test user default suiteName in AppCoordinatorTests

dvdchr committed:
Add more tests for RoleErrorViewModel

dvdchr committed:
Fix AppCoordinatorTests

dvdchr committed:
Merge branch 'task/3963-role-eligibility-use-case' into task/3963-role-check-flow

dvdchr committed:
Merge branch 'task/3963-role-eligibility-use-case' into task/3963-role-check-flow

Ecarrion commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerCoordinator.swift:
```diff
@@ -45,17 +49,19 @@ final class StorePickerCoordinator: Coordinator {
+                // if user is eligible, then switch to the desired store.
+                self.roleEligibilityUseCase.reset()
+                self.switchStore(with: storeID, onCompletion: onCompletion)
+                return
```
⚠️  I think it looks a bit strange to handle the `success` case in a `guard let`.
What do you think about returning a `Result` from `roleEligibilityUseCase.checkEligibility` and handle these paths in a switch?

dvdchr replied:
Thanks for the suggestion! I've received [similar review in Part 3](https://github.com/woocommerce/woocommerce-ios/pull/4492#discussion_r662318120), and I'll be changing this to a `Result<Void, Error>` soon! 🙂

dvdchr replied:
Update: comment addressed in #4492, particularly in [this commit](https://github.com/woocommerce/woocommerce-ios/pull/4492/commits/f2e1cccf92c216a97dcf4ab30806c9024fd2330e)! 🙂

Ecarrion commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerCoordinator.swift:
```diff
@@ -85,4 +91,43 @@ private extension StorePickerCoordinator {
         }
     }
 
+    func switchStore(with storeID: Int64, onCompletion: @escaping SelectStoreClosure) {
```
nit: Some documentation here will also be nice!

dvdchr replied:
Will do! 👍

Ecarrion commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerCoordinator.swift:
```diff
@@ -85,4 +91,43 @@ private extension StorePickerCoordinator {
+        while let presented = topNavigationController.presentedViewController as? UINavigationController {
+            topNavigationController = presented
+        }
+        topNavigationController.pushViewController(errorViewController, animated: true)
```
nit: 
```suggestion
        topNavigationController.show(errorViewController, sender: self)
```

dvdchr replied:
Good catch! Will adjust this soon.

Ecarrion commented on WooCommerce/Classes/Authentication/Navigation Exceptions/RoleErrorViewModel.swift:
```diff
@@ -14,6 +17,10 @@ final class RoleErrorViewModel {
         .incorrectRoleError
     }
 
+    var onSuccess: (() -> Void)?
+
+    var onDeauthenticationRequest: (() -> Void)?
```
Since this is an API meant to be consumed It would be good to add some docs too!

dvdchr replied:
Noted, will add some docs soon! 👍

Ecarrion commented on WooCommerce/Classes/Authentication/Navigation Exceptions/RoleErrorViewModel.swift:
```diff
@@ -38,17 +45,37 @@ final class RoleErrorViewModel {
+        roleEligibilityUseCase.checkEligibility(for: siteID) { [weak self] error in
+            guard let self = self else { return }
+
+            guard let error = error else {
+                self.roleEligibilityUseCase.reset()
+                self.onSuccess?()
+                return
+            }
+
+            if case let RoleEligibilityError.insufficientRole(errorInfo) = error {
+                self.titleText = errorInfo.name
+                self.subtitleText = errorInfo.humanizedRoles
+                self.output?.refreshTitleLabels()
+                self.output?.displayNotice(message: .insufficientRolesErrorMessage)
+                return
+            }
+
+            self.output?.displayNotice(message: .retrieveErrorMessage)
```
Same comment as above!

Ecarrion commented on WooCommerce/Classes/Authentication/RoleEligibilityUseCase.swift:
```diff
@@ -35,8 +35,11 @@ final class RoleEligibilityUseCase {
     private var lastErrorInfo: EligibilityErrorInfo? {
-        didSet {
-            defaults.setValue(lastErrorInfo?.toDictionary(), forKey: Constants.eligibilityErrorInfoKey)
+        get {
+            loadLastErrorInfo()
+        }
+        set {
+            defaults.setValue(newValue?.toDictionary(), forKey: Constants.eligibilityErrorInfoKey)
```
⚠️  We try to not save more things to user defaults. For settings like these, we have `GeneralAppSettings` which is accessible through a `Yosemite` store.

dvdchr replied:
Noted. Will look into this and move the implementation to `GeneralAppSettings` instead. 👍

dvdchr replied:
Addressed in #4665 !

Ecarrion commented on WooCommerce/Classes/ViewRelated/AppCoordinator.swift:
```diff
@@ -66,6 +69,14 @@ private extension AppCoordinator {
     func displayLoggedInUI() {
+        // if the previous role check indicates that the user is ineligible, then we want to show the error message.
+        // Eligibility status is stored locally in user defaults, and updated asynchronously upon app launch.
+        if let errorInfo = roleEligibilityUseCase.lastEligibilityErrorInfo(),
```
⚠️  I think this is a bit missleading. A use case is supposed to serve a single purpose, but this use-case already has two. `checkEligibility` and `lastEligibilityErrorInfo`. What do you think about moving `lastEligibilityErrorInfo()` to `AppSettingsStore`?

dvdchr replied:
> A use case is supposed to serve a single purpose,

Ah, I didn't know that we are supposed to have only one public method per use case. I thought it's more of being responsible for a single context or object, but still allowing more than one method. I'll try to separate the responsibility to another object, then 👍 

> moving lastEligibilityErrorInfo() to AppSettingsStore?

Good idea! Will do this soon. 👍

dvdchr replied:
Update: comment addressed in #4492, particularly in [this commit](https://github.com/woocommerce/woocommerce-ios/pull/4492/commits/f2e1cccf92c216a97dcf4ab30806c9024fd2330e)! 🙂

dvdchr replied:
This is addressed in #4665!

Ecarrion commented on WooCommerce/Classes/ViewRelated/AppCoordinator.swift:
```diff
@@ -90,6 +101,25 @@ private extension AppCoordinator {
+        }
+
+        // this needs to be wrapped within a navigation controller to properly show the right bar button for Help.
+        setWindowRootViewControllerAndAnimateIfNeeded(UINavigationController(rootViewController: errorViewController))
```
⚠️  I think we should use `WooNavigationController` to be safe with any future global customization.

dvdchr replied:
Ah, right! About this, I tried using `WooNavigationController` but it was [overriding the status bar style](https://github.com/woocommerce/woocommerce-ios/blob/develop/WooCommerce/Classes/System/WooNavigationController.swift#L25-L30) to `.lightContent` regardless of the current color scheme. 

I'll follow your suggestion on using `WooNavigationController` for now, and will probably raise this as a separate issue later. Thanks! 👍

Ecarrion reviewed (changes requested):
Hi @dvdchr! Thanks for tackling this. I left some concerns marked with ⚠️ 

----

I'm not sure if this is already on your radar but we are missing some accessibility support. See how the feature looks when using big fonts:

Portrait | Landscape
---- | ----
<img width="381" alt="problem-big-fonts" src="https://user-images.githubusercontent.com/562080/124191549-8a43b580-da89-11eb-9801-b47e90337979.png"> | <img width="809" alt="problem-big-fonts-2" src="https://user-images.githubusercontent.com/562080/124191547-89128880-da89-11eb-9e98-6741a266baa8.png">

----

Also I think it would be a great improvement to always do the async check when the app launches and act immediately, not wait until the next app launch or until the user taps retry. 


https://user-images.githubusercontent.com/562080/124191734-d8f14f80-da89-11eb-8bdb-377b9ad7e9fc.mov

dvdchr commented:
> we are missing some accessibility support

Thanks for noticing this. I missed testing large font sizes. Since UI is addressed in Part 2, I'll address this in a different PR!

> it would be a great improvement to always do the async check when the app launches and act immediately

I see, thanks for the suggestion 👍 . Previously, the consideration for delaying the error page to next launch was to not disrupt the user abruptly – but if their roles are ineligible, the app would not be working correctly for them and it would confuse them anyways. To make this work, I'll probably hook into the `AppCoordinator` and use `CombineLatest3` to listen to eligibility state. When the state changes, it can be immediately acted upon.

dvdchr commented:
> we are missing some accessibility support. 

This will be resolved in #4619!

shiki commented:
Hi, @dvdchr and @Ecarrion! Would you mind marking the conversations that have been resolved as "Resolved"? Just so we have some idea on what else is pending and how we can help each other get this PR merged. 🙂 

Is this ready for another round of review? 🙂

dvdchr committed:
Add docs for switchStore method

dvdchr committed:
Add some minor changes

- Add some comments
- Use WooNavigationController to wrap the role error page

dvdchr committed:
Merge branch 'task/3963-role-eligibility-use-case' into task/3963-role-check-flow

dvdchr committed:
Adjust for Result type changes, remove reset() method from use case

dvdchr committed:
minor: move private method to private extension

dvdchr committed:
Fix tests

dvdchr committed:
Fix RoleErrorViewModel tests

dvdchr committed:
Merge branch 'task/4492-eligibility-info-appsettings' into task/3963-role-check-flow

dvdchr committed:
Refactor AppCoordinator and AppDelegate to react faster on ineligible status

dvdchr committed:
minor: Adjust EligibilityErrorInfo usages to StorageEligibilityErrorInfo

dvdchr committed:
Adjust tests

dvdchr commented:
Hey there @shiki! Thanks for checking in. 🙂 

> Is this ready for another round of review? 

While this PR is still dependent on pending PRs, it is now at a working state. Here's an update based on the latest major review comments from @Ecarrion :

- Missing some accessibility support: addressed in #4619 ✅ 
- Always do the async check when the app launches and act immediately: addressed in https://github.com/woocommerce/woocommerce-ios/pull/4493/commits/9b0cae812786f56d5e16659f794ab261c339ba88 ✅ 
- Move `EligibilityErrorInfo` model to `Storage`, and expose it through `AppSettingsStore`: addressed in #4665 (⏳ pending review)
- All the other minor review comments are addressed here (marked as resolved). ✅ 

---
Here's a chronological list on which PR should be reviewed first:

- #4620. This is an integration branch, so it is safe to merge to `develop`.
- #4665. Since this targets [Part 3](https://github.com/woocommerce/woocommerce-ios/pull/4492), this should be reviewed first.
- #4492 
- And finally this PR!

dvdchr committed:
Fix tests

dvdchr committed:
Merge branch 'task/4492-eligibility-info-appsettings' into task/3963-role-check-flow

dvdchr committed:
Merge branch 'task/4492-eligibility-info-appsettings' into task/3963-role-check-flow

dvdchr committed:
Clean up user defaults dependency from app coordinator tests

Ecarrion reviewed (approved):
Thanks for the changes @dvdchr!
Code and tests look good to me! 👏

dvdchr committed:
Add a release note
